### ISSUE TITLE
Training parts of an equinox module

### ISSUE THREAD

**Issue by ParticularlyPythonicBS** [patrick-kidger/equinox#406]
Firstly, thank you for creating and maintaining this incredibly useful work.

I am trying to use equinox alongside diffrax and optax for neural network (optimal) control.
To do this I have a series of modules and submodules
```
class Dynamics(eqx.Module) # no trainable parameter, a physical system that returns dx

class TimeController(eqx.module) # simple MLP

class Controlled_dynamics(eqx.Module): # combines Dynamics and Time Controller
   Dynamics: eqx.Module
    Controller: eqx.Module
    def __init__(self, Dynamics, Controller):
        self.Dynamics = Dynamics
        self.Controller = Controller

    def __call__(self, t, y, args):
        u = self.Controller(t, y)
        dx = self.Dynamics(t, y, u)
        return dx

class NeuralODE(eqx.Module): # takes the controlled dynamics and uses diffrax to solve the system
    Controlled_dynamics: eqx.Module

    def __init__(self, Controlled_dynamics):
        super().__init__()
        self.Controlled_dynamics = Controlled_dynamics

    def __call__(self, ts, y0):
        terms = diffrax.ODETerm(self.Controlled_dynamics)
        solver = diffrax.Tsit5()
        t0 = ts[0]
        t1 = ts[-1]
        dt0 = ts[1] - ts[0]
        saveat = diffrax.SaveAt(ts=ts)
        stepsize_controller = diffrax.PIDController(rtol=1e-3, atol=1e-6)
        adjoint_controller = diffrax.PIDController(rtol=1e-3, atol=1e-6, norm=diffrax.adjoint_rms_seminorm)
        adjoint = diffrax.BacksolveAdjoint(stepsize_controller=adjoint_controller)
        sol = diffrax.diffeqsolve(terms,
                        solver,
                        t0,
                        t1,
                        dt0,
                        y0,
                        saveat=saveat, stepsize_controller=stepsize_controller,
                        adjoint=adjoint)
        return sol.ys
```
During training, I need to pass objects of neural_ode since my loss depends on the solution, but the only trainable components are in the time controller.
Naively performing the update by,
```
@eqx.filter_jit
def make_step(model, x, targets, ts, optim, opt_state):
    loss, grads = compute_loss(model, x, targets, ts)
    updates, opt_state = optim.update(grads, opt_state)
    model = eqx.apply_updates(model, updates)
    return loss, model, opt_state
```
does not seem to train the model or change the loss at all. 

I would appreciate any advice on fixing what parameters are trainable for such a system.

Thanks a lot!

**Comment by ParticularlyPythonicBS** [comment]
This works actually, I just had a bug in my training code.
